sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/InsertIntoHadoopFsRelationCommand.scala

```diff
@@ -266,4 +304,131 @@ case class InsertIntoHadoopFsRelationCommand(
+              staticPartitionKVs.slice(0, i))
+            val checkedPath = if (!subPath.isEmpty) {
+              new Path(stagingDir, subPath)
+            } else {
+              stagingDir
+            }
+            if (fs.exists(checkedPath)) {
+              throwConflictException(fs, stagingDir, i, staticPartitionKVs)
+            }
+          }
+        }
+      }
+    }
+  }
+
+  private def throwConflictException(
+      fs: FileSystem,
+      stagingDir: Path,
+      depth: Int,
+      staticPartitionKVs: Seq[(String, String)]): Unit = {
+    val conflictedPaths = ListBuffer[Path]()
+    val currentPath = if (depth == staticPartitionKVs.size || staticPartitionKVs.size == 0) {
+      stagingDir
+    } else {
+      new Path(stagingDir, HadoopMapReduceCommitProtocol.getEscapedStaticPartitionPath(
+        staticPartitionKVs.slice(0, staticPartitionKVs.size - depth)))
+    }
+
+    findConflictedStagingOutputPaths(fs, currentPath, depth, conflictedPaths)
+
+    val pathsInfo = conflictedPaths.toList
+      .map { path =>
+        val absolutePath = path.toUri.getRawPath
+        val relativePath = absolutePath.substring(absolutePath.lastIndexOf(stagingDir.getName))
+        var appId: Option[String] = None
+        var modificationTime: Date = null
+        try {
+          val files = fs.listStatus(path)
+          if (files.size > 0) {
+            appId = Some(files.apply(0).getPath.getName)
+          }
+          modificationTime = new Date(fs.getFileStatus(path).getModificationTime)
+        } catch {
+          case e: Exception => logWarning("Exception occurred", e)
+        }
+        (relativePath, appId.getOrElse("Not Found"), modificationTime)
+      }
+
+    throw new InsertFileSourceConflictException(
+      s"""
+         | Conflict is detected, some other conflicted output path(s) under tablePath:
+         | ($outputPath) existed.
+         | Relative path, appId and last modification time information is shown as below:
+         | ${pathsInfo}.
+         | There may be two possibilities:
+         | 1. Another InsertDataSource operation is executing, you need wait for it to
+         |    complete.
+         | 2. This dir is belong to a killed application and not be cleaned up gracefully.
+         |
+         | Please check the last modification time and use given appId to judge whether
+         | relative application is running now. If not, you should delete responding path
+         | without recursive manually.
+         |""".stripMargin)
+  }
+
+  /**
+   * Find relative staging output paths, which is conflicted with current
+   * InsertIntoHadoopFsRelation operation.
+   */
+  private def findConflictedStagingOutputPaths(
+      fs: FileSystem,
+      path: Path,
+      depth: Int,
+      paths: ListBuffer[Path]): Unit = {
+    try {
+      if (fs.exists(path)) {
+        if (depth == 0) {
+          paths += path
+        } else {
+          for (file <- fs.listStatus(path)) {
```

Comment:
listStatus will raise an FNFE if the path isn't there. So you could optimise away that previous exists check if depth != 0

Reply:
I checked whether this path is existed firstly and caught all exceptions in this code block.